File: tweet/worker.py
import configparser
import pymongo
import sys, os
import math, time
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from protocols.schema import *
from protocols.messages import *
from protocols.rpc_protocols import *

# Get consts from config file
RES_SUCCESS = STATUS_OK
RES_FAILURE = STATUS_ERROR

config = configparser.ConfigParser()
config.read('./config.ini')

config_tweet = config['TWEET']
SEARCH_LIMIT_DEFAULT = int(config_tweet['Search_Limit_Default'])
SEARCH_LIMIT_MAX = int(config_tweet['Search_Limit_Max'])
DB_NAME = config_tweet['MongoDB_Name']
NODE_NAME = config_tweet['MongoDB_Node']
PORT_NUM = int(config_tweet['MongoDB_Mgs_Port'])
TWEET_COLLECTION_NAME = config_tweet['MongoDB_Tweet_Collection']
PROFILE_COLLECTION_NAME = config_tweet['MongoDB_Profile_Collection']

# Set up Mongo client
client = pymongo.MongoClient('mongodb://%s' % NODE_NAME, PORT_NUM, maxPoolSize=100, waitQueueMultiple=10)


def add_tweet(payload):
    payload['childType'] = None if 'childType' not in payload else payload['childType']
    payload['parent'] = None if 'parent' not in payload else payload['parent']
    payload['media'] = [] if 'media' not in payload else payload['media']

    tweet = json.loads(new_post(payload['username'], payload['content'], 
        payload['childType'], payload['parent'], payload['media']))
    tweet_id = tweet['id']

    db = client[DB_NAME]
    collection = db[TWEET_COLLECTION_NAME]
    collection.create_index([('timestamp', pymongo.DESCENDING), 
        ('id', pymongo.ASCENDING),
        ('username', pymongo.ASCENDING)], 
        background=True)
    result = collection.insert_one(tweet)

    if(result.inserted_id == None):
        return generate_message(RES_FAILURE, ERROR_POST_TWEET)

    res = json.dumps({
        'status': RES_SUCCESS,
        'id': tweet_id,
        'item': tweet
    })
    return res


def get_tweet(payload):
    query = json.loads(tweet_query(payload['id']))

    db = client[DB_NAME]
    collection = db[TWEET_COLLECTION_NAME]
    result = collection.find_one(query)

    if(result == None):
        return generate_message(RES_FAILURE, ERROR_GET_TWEET)
    del result['_id']
    
    res = json.dumps({
        'status': RES_SUCCESS,
        'item': result
    })
    return res


def delete_tweet(payload):
    query = json.loads(user_tweet_query(payload['id'], payload['username']))

    db = client[DB_NAME]
    collection = db[TWEET_COLLECTION_NAME]
    tweet = collection.find_one(query)
    result = collection.delete_one(query)

    if(result.deleted_count != 1):
        return generate_message(RES_FAILURE, ERROR_GET_TWEET)
    return generate_del_msg(RES_SUCCESS, '', tweet['media'] if tweet != None else None)


def like_tweet(payload):
    like = True if 'like' not in payload else payload['like']
    
    query = json.loads(tweet_query(payload['id']))
    user = payload['username']

    db = client[DB_NAME]
    collection = db[TWEET_COLLECTION_NAME]
    result = collection.find_one(query)
    if result == None:
        return generate_message(RES_FAILURE, ERROR_GET_TWEET)
    liked_by = result['property']['liked_by']
    
    if like:
        if user in liked_by:
            return generate_message(RES_SUCCESS, '')
        else:
            update = json.loads(like_tweet_update(user))
            collection.update_one(query, update, upsert=False)
    else:
        if user in liked_by:
            update = json.loads(unlike_tweet_update(user))
            collection.update_one(query, update, upsert=False)
        else:
            return generate_message(RES_SUCCESS, '')


def search(payload):
    db = client[DB_NAME]
    tweet_collection = db[TWEET_COLLECTION_NAME]
    profile_collection = db[PROFILE_COLLECTION_NAME]

    # Process queries conditions
    timestamp = math.floor(time.time()) if 'timestamp' not in payload else int(payload['timestamp'])
    q = None if 'q' not in payload else payload['q']
    user = None if 'user' not in payload else payload['user']
    username = None if 'username' not in payload else payload['username']
    following = True if 'following' not in payload else payload['following']
    rank = 'interest' if 'rank' not in payload else payload['rank']
    parent = None if 'parent' not in payload else payload['parent']
    replies = True if 'replies' not in payload else payload['replies']
    hasMedia = False if 'hasMedia' not in payload else payload['hasMedia']
    targets = None
    if following:
        targets = []
        result = profile_collection.find_one({'username': user})
        if result != None:
            targets = result['following']
    limit = SEARCH_LIMIT_DEFAULT
    if 'limit' in payload:
        limit = int(payload['limit']) if int(payload['limit']) < SEARCH_LIMIT_MAX else SEARCH_LIMIT_MAX

    # Start Query
    query = json.loads(search_query(timestamp, q, username, targets, parent, replies, hasMedia))

    #sort by interest or time
    cursor = tweet_collection.find(query).sort(search_sort(rank)).limit(limit)
    if(cursor == None):
        res = json.dumps({
            'status': RES_SUCCESS,
            'items': []
        })
        return res

    result = []
    for doc in cursor:
        del doc['_id']
        result.append(doc)
    
    res = json.dumps({
        'status': RES_SUCCESS,
        'items': result
    })
    
    return res

File: tweet/worker.js
const mongodb = require('mongodb').MongoClient;
const utils = require('./protocols/utils');
const mongoose = require('mongoose');
const mexp = require('mongoose-elasticsearch-xp');

const MONGO_URI = require('./config').tweet.MongoDB_Uri;
const DB_NAME = require('./config').tweet.MongoDB_Name;
//const TWEET_COLLECTION = require('./config').tweet.MongoDB_Tweet_Collection;
const PROFILE_COLLECTION = require('./config').tweet.MongoDB_Profile_Collection;
const ES_HOST = require('./config').tweet.ES_HOST;
const ERROR_GET_TWEET = require('./protocols/messages').ERROR_GET_TWEET;
const ERROR_POST_TWEET = require('./protocols/messages').ERROR_POST_TWEET;
const ERROR_DELETE_TWEET = require('./protocols/messages').ERROR_DELETE_TWEET;

const STATUS_OK = 'OK';
const STATUS_ERROR = 'error';

// Initialize mongo connection
var db = null;
mongodb.connect(MONGO_URI, function(err, client) {
    if(err) throw err;
    db = client.db(DB_NAME);
});

// Initialize mongoose connection
mongoose.connect(MONGO_URI, {
    dbName: DB_NAME,
    autoIndex: false,
    reconnectTries: Number.MAX_VALUE,
    reconnectInterval: 500,
    bufferMaxEntries: 0
});

var tweetSchema = utils.getTweetSchema();
tweetSchema.plugin(mexp, {hosts: [ES_HOST]});
if (!tweetSchema.options.toObject) tweetSchema.options.toObject = {};
tweetSchema.options.toObject.transform = function (doc, ret, options) {
    // remove the auto generated value of every document before returning the result
    delete ret._id;
    delete ret.__v;
    return ret;
}

var Tweet = mongoose.model('tweet', tweetSchema);


exports.addTweet = function(payload) {
    return new Promise(function(resolve, reject) {
        var tweet = new Tweet(utils.tweetInsert(payload.id, payload.username, payload.content,  
            payload.childType, payload.parent, payload.media));
        tweet.save(function (err) {
            if(err) {
                console.error(err.message);
                return;
            }

            var response = {
                'status': STATUS_OK,
                'id': payload.id,
                'item': tweet.toObject()
            };
            resolve(response);
        });
    });
}

exports.getTweet = function(payload) {
    return new Promise(function(resolve, reject) {
        var query = utils.tweetQuery(payload.id);
        
        Tweet.findOne(query).lean().exec(function(err, result) {
            if(err) {
                resolve(utils.generateMessage(STATUS_ERROR, err.message));
                return;
            }

            if(result != null) {
                delete result._id;
                delete result.__v;

                var response = {
                    'status': STATUS_OK,
                    'item': result
                };
                resolve(response);
            }
            else {
                resolve(utils.generateMessage(STATUS_ERROR, ERROR_GET_TWEET));
            }
        });   
    });
}

exports.deleteTweet = function(payload) {
    return new Promise(function(resolve, reject) {
        var query = utils.tweetQueryWithUsername(payload.id, payload.username);

        Tweet.findOneAndRemove(query, function(err, result) {
            if(err) {
                resolve(utils.generateMessage(STATUS_ERROR, err.message));
                return;
            }

            if(result != null) {
                var response = {
                    'status': STATUS_OK,
                    'media': result.toObject().media
                };
                resolve(response);
            }
            else {
                resolve(utils.generateMessage(STATUS_ERROR, ERROR_DELETE_TWEET));
            }
        });         
    });
}

exports.likeTweet = function(payload) {
    return new Promise(function(resolve, reject) {
        var query = utils.tweetQuery(payload.id);

        Tweet.findOne(query).lean().exec(function(err, result) {
            if(err) {
                resolve(utils.generateMessage(STATUS_ERROR, err.message));
                return;
            }

            if(result != null) {
                var user = payload.username;
                var likedBy = result.property.liked_by;
                var update = null;

                if(payload.like && !likedBy.includes(user)) {
                    update = utils.likeTweetUpdate(user);
                }
                else if(!payload.like && likedBy.includes(user)) {
                    update = utils.unlikeTweetUpdate(user);
                }

                if(update != null) {
                    Tweet.findOneAndUpdate(query, update, function(err, result) {
                        if(err) {
                            resolve(utils.generateMessage(STATUS_ERROR, err.message));
                            return;
                        }
                    });
                }
                resolve(utils.generateMessage(STATUS_OK, ''));
            }
            else {
                resolve(utils.generateMessage(STATUS_ERROR, ERROR_GET_TWEET));
            }
        });     
    });
}

exports.searchTweet = function(payload) {
    return new Promise(function(resolve, reject) {
        db.collection(PROFILE_COLLECTION).findOne({'username': payload.username}, function(err, result) {
            if(err) {
                resolve(utils.generateMessage(STATUS_ERROR, err.message));
                return;
            }

            // Resolve list of following users if is logged in
            if(result != null && payload.following) {
                payload.targets = result.following;
            }
            else if(result == null && payload.following) {
                payload.targets = [];
            }

            var query = utils.searchQuery(payload.limit, parseInt(payload.timestamp), payload.q, payload.target, 
                payload.targets, payload.parent, payload.replies, payload.hasMedia, payload.rank);

            Tweet.esSearch(query).then(function(results) {
                var response = {
                    'status': STATUS_OK,
                    'items': []
                };
                console.log("result: " + results);

                if(results != null) {
                    for(var i = 0; i < results.length; i++) {
                        delete results[i]._id;
                        response.items.push(results[i]);
                    }
                }
                resolve(response);
            });      
        });            
    });
}